### DISCUSSION TITLE
Can I disable HPA while the argo rollout is progressing?

### DISCUSSION THREAD

**Discussion by sungho-rpls** [argoproj/argo-rollouts#3925] [Q&A]
HPA scales out replicas in my environment because of high CPU usage while the argo rollout progresses. High CPU usage for startup time is acceptable for me, but too many replicas by HPA cause problems with latency, resource usage, and database connections. Is it possible to disable HPA scaling when the argo rollout is progressing?

**Comment by kostis-codefresh** [comment]
If you are doing blue/green deployment use the `previewReplicaCount` replica to lock down preview pods.

If you are doing canary deployments use `setCanaryScale` to lock down canary pods.

Would that work for you?

**Reply by florianmutter** [reply]
We are facing the same challenge. We have a lot of different workloads where some only run 2 pods and others run 20 or more. Having to come up with a good value for `setCanaryScale` seems hard. We would like to use the rollout feature of adjusting the pod count according to traffic split. 

Right now this does not work at all because the pods do require more cpu on startup. This causes the HPA to start scaling after rollout created the first set of canary pods which causes rollout to adjust the canary replica count which causes even more cpu load.

Would it be somehow possible to make the HPA only look at the stable replicaset and ignore the canary one?
Another option could be to make rollout behave more like deployment rolling update: Only start one canary pod, wait until it is ready, shutdown a stable pod, repeat until desired number of canary pods is reached, start analysis,...

**Reply by kostis-codefresh** [reply]
> Another option could be to make rollout behave more like deployment rolling update

This can be done with the `dynamicStableScale` Option

See example 07 here https://github.com/kostis-codefresh/rollouts-autoscaling-example

**Reply by GurayCetin** [reply]
> We are facing the same challenge. We have a lot of different workloads where some only run 2 pods and others run 20 or more. Having to come up with a good value for `setCanaryScale` seems hard. We would like to use the rollout feature of adjusting the pod count according to traffic split.
> 
> Right now this does not work at all because the pods do require more cpu on startup. This causes the HPA to start scaling after rollout created the first set of canary pods which causes rollout to adjust the canary replica count which causes even more cpu load.
> 
> Would it be somehow possible to make the HPA only look at the stable replicaset and ignore the canary one? Another option could be to make rollout behave more like deployment rolling update: Only start one canary pod, wait until it is ready, shutdown a stable pod, repeat until desired number of canary pods is reached, start analysis,...

As a follow-up to the previous case, we have applied the `dynamicStableScale` option. While new pods are being added to the canary, we observed that stable pods are being terminated simultaneously. However the HPA is actively scaling both canary and stable ReplicaSets, we believe it is scaling more aggressively than expected.

In one instance, the rollout began with 21 pods. By the end of the analysis phase, the total number of pods had increased to 41 — meaning 20 additional pods were created, which were later scaled down to the optimal number.

This scaling behavior leads to unnecessary resource consumption during rollouts, as it also triggers an increase in node count to support the temporary surge in pods.

**Reply by kostis-codefresh** [reply]
You can use previewReplicaCount to lock down the number of preview/unstable pods. And then HPA will just apply on stable pods. 

Would that work in your case?

**Reply by GurayCetin** [reply]
I guess `previewReplicaCount` is valid for blue-green strategy not canary, and we do not prefer fixed values considering our various pod numbers 
https://argo-rollouts.readthedocs.io/en/stable/features/specification/

**Reply by kostis-codefresh** [reply]
Yes setCanaryScale is the same option for Canaries.

https://github.com/kostis-codefresh/rollouts-autoscaling-example/blob/main/10-hpa-canary-decoupled/rollout.yaml#L28

**Reply by kostis-codefresh** [reply]
The problem with this discussion is that different people have different needs. Some people consider a deployment a special case and don't want HPA running/affecting during rollouts and some people want HPA to always take effect (even when a canary is in progress).

> we believe it is scaling more aggressively than expected

Maybe you can fine-tune the HPA configuration to be less aggressive, by taking into account canary deployment as well (and not just the normal runtime periods)? Where do HPA now gets its metrics from? Memory? CPU? Custom metrics?

**Reply by florianmutter** [reply]
We basically want HPA and we cannot use `setCanaryScale` because the rollout might happen during high and low traffic times. What we noticed is: If we just use deployment rolling update we do not see the same scaling behavior as we see with rollout. We want Rollout to behave like deployment rolling update regarding the number of pods started and stopped. We have set `maxSurge` to 1 but it seems this is not respected during rollout.

**Reply by florianmutter** [reply]
Possibly related:
* https://github.com/argoproj/argo-rollouts/issues/2239
* https://github.com/argoproj/argo-rollouts/issues/3284
* https://github.com/argoproj/argo-rollouts/issues/4260

**Reply by florianmutter** [reply]
Here is an observation of ours: It seems that shifting traffic with istio does not happen as fast as stable pods are shut down if `dynamicStableScale` is enabled. You can see in the following graph that CPU usage of the remaining stable pods doubles because rollout shuts half of them down (weight set to 50%) before the traffic was shifted to the canary pods.

Blue line is number of stable pods, yellow line number of canary pods. The black line marks where canary pods are ready and stable pods are shut down. CPU usage increases because 90% of traffic is still routed to stable pods. This could be something that does contribute to the HPA scaling up 1 minute later.  

![Screenshot 2025-07-03 at 13 54 21](https://github.com/user-attachments/assets/30ea7383-ed8d-4c88-815c-96cea6a18b0f)

Is there a way to make rollout scale down stable slower?

**Reply by kostis-codefresh** [reply]
But is this a problem of Argo Rollouts or Istio?

What does your Rollout look like? Do you have separate setCanaryScale/setWeight directives?

**Reply by GurayCetin** [reply]
As three steps; setWeight %10 -> %50 -> %80 
```
spec:
  strategy:
    canary:
      steps:
        - setWeight: 10
        - pause: {duration: 5m}
        - analysis:
            templates:
              - templateName: datadog-error-rate
                clusterScope: true
        - setWeight: 50
        - pause: {duration: 5m}
        - analysis:
            templates:
              - templateName: datadog-error-rate
                clusterScope: true
        - setWeight: 80
        - pause: {duration: 5m}
        - analysis:
            templates:
              - templateName: datadog-error-rate
                clusterScope: true
```

**Reply by rave2101** [reply]
@GurayCetin were you able to find any work around for this issue?
